dipy/reconst/mcsd.py

```diff
@@ -0,0 +1,347 @@
+        self.m = np.zeros_like(self.n)
+        self.shells = shells
+        if self.iso < 1:
+            raise ValueError("sh_order and shape of response do not agree")
+
+    @property
+    def iso(self):
+        return self.response.shape[1] - (self.sh_order // 2) - 1
+
+
+def closest(haystack, needle):
+    diff = abs(haystack[:, None] - needle)
+    return diff.argmin(axis=0)
+
+
+def _inflate_response(response, gtab, n, delta):
+    if any((n % 2) != 0) or (n.max() // 2) >= response.sh_order:
+        raise ValueError("Response and n do not match")
+
+    iso = response.iso
+    n_idx = np.empty(len(n) + iso, dtype=int)
+    n_idx[:iso] = np.arange(0, iso)
+    n_idx[iso:] = n // 2 + iso
+
+    b_idx = closest(response.shells, gtab.bvals)
+    kernal = response.response / delta
+
+    return kernal[np.ix_(b_idx, n_idx)]
+
+
+def _basic_delta(iso, m, n, theta, phi):
+    """Simple delta function
+    Parameters
+    ----------
+    iso: int (optional)
+        Number of tissue compartments for running the MSMT-CSD. Minimum
+        number of compartments required is 2.
+        Default: 2
+    m : int ``|m| <= n``
+        The order of the harmonic.
+    n : int ``>= 0``
+        The degree of the harmonic.
+    theta : array_like
+       inclination or polar angle
+    phi : array_like
+       azimuth angle
+    """
+    wm_d = shm.gen_dirac(m, n, theta, phi)
+    iso_d = [SH_CONST] * iso
+    return np.concatenate([iso_d, wm_d])
+
+
+def _pos_constrained_delta(iso, m, n, theta, phi, reg_sphere=default_sphere):
+    """
+    Delta function optimized to avoid negative lobes. Implements a Linear
+    Programming solver from `CVXPY` to impose this positivity constraint. The
+    default solver used is GLPK.
+
+    Parameters
+    ----------
+    iso: int (optional)
+        Number of tissue compartments for running the MSMT-CSD. Minimum
+        number of compartments required is 2.
+        Default: 2
+    m : int ``|m| <= n``
+        The order of the harmonic.
+    n : int ``>= 0``
+        The degree of the harmonic.
+    theta : array_like
+       inclination or polar angle
+    phi : array_like
+       azimuth angle
+    reg_sphere : Sphere (optional)
+        sphere used to build the regularization B matrix.
+        Default: 'symmetric362'.
+        weight given to the constrained-positivity regularization part of
+        the deconvolution equation (see [1]_). Default: 1
+    """
+
+    x, y, z = geo.sphere2cart(1., theta, phi)
+
+    # Realign reg_sphere so that the first vertex is aligned with delta
+    # orientation (theta, phi).
+    M = geo.vec2vec_rotmat(reg_sphere.vertices[0], [x, y, z])
+    new_vertices = np.dot(reg_sphere.vertices, M.T)
+    _, t, p = geo.cart2sphere(*new_vertices.T)
+
+    B = shm.real_sph_harm(m, n, t[:, None], p[:, None])
+    G_temp = np.ascontiguousarray(B[:, n != 0])
+    # c_ samples the delta function at the delta orientation.
+    c_temp = G_temp[0][:, None]
```

Comment:
Why are we pulling out the first element of this array?

Reply:
That would  be a question for @MrBago ... it works as expected,  nevertheless :)